mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,929 @@
+    // Returns a u64 slice that allows access L1 page entries represented by pfn.
+    fn l1_slice_from_pfn(pfn: SupervisorPfn) -> &'a mut [u64] {
+        let l1_page_addr = (pfn.bits() << 12) as *mut MttL1Page;
+        // Safety: We are creating a reference to page based on a PFN in the MTT L2 table
+        // The PFN entry must have been preallocated by trusted platform-FW, or was created
+        // from a PageList that was verified during PlatformMttTables::init()
+        let l1_page = unsafe { &mut *core::ptr::slice_from_raw_parts_mut(l1_page_addr, 1) };
+        &mut l1_page[0].base[0..]
+    }
+}
+
+struct PlatformMttInfo {
+    mtt_l2_base: &'static mut [u64],
+    l1_page_list: PageList<Page<InternalClean>>,
+}
+
+/// Holds references to the platform allocated L2 and L2 tables in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2 and L1 tables.
+pub struct PlatformMttTables {
+    inner: Mutex<PlatformMttInfo>,
+}
+
+impl PlatformMttTables {
+    // Performs sanity checks on the platform L2 table
+    fn validate_mtt(&self) -> Result<()> {
+        let mtt_info = self.inner.lock();
+        let l2_base_addr = mtt_info.mtt_l2_base.as_ptr() as usize;
+        // L2 must be 8MB aligned with a length of 8MB
+        let is_l2_aligned = l2_base_addr & (L2_BYTE_SIZE - 1) == 0;
+        let is_l2_size_ok = mtt_info.mtt_l2_base.len() == L2_BYTE_SIZE / 8;
+        // TODO: Add additional checks to validate individual L2 entries
+        let addr = SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr as u64))
+            .ok_or(InvalidAddress)?;
+        if is_l2_aligned && is_l2_size_ok {
+            let mut mtt_range = MttRange::new(addr, L2_BYTE_SIZE, mtt_info)?;
+            while let Some((addr, region_size)) = mtt_range.next() {
+                let region_memory_type = mtt_range.get_page_range_memory_type(addr, region_size)?;
+                if region_memory_type != MttMemoryType::Confidential {
+                    return Err(InvalidPlatformConfiguration);
+                }
+            }
+        } else {
+            return Err(InvalidPlatformConfiguration);
+        }
+
+        Ok(())
+    }
+
+    // Test only NOP validation
+    #[cfg(test)]
+    fn validate_mtt(&self) -> Result<()> {
+        let mtt_info = self.inner.lock();
+        // Intended only to silence clippy warning about unused imports while testing
+        let l2_base_addr = mtt_info.mtt_l2_base.as_ptr() as u64;
+        SupervisorPageAddr::new(RawAddr::supervisor(l2_base_addr)).ok_or(InvalidAddress)?;
+        Ok(())
+    }
+
+    /// Creates the singleton Mtt instance.
+    /// # Safety
+    ///
+    /// The caller must ensure that:
+    /// 1: `mtt_l2_base` is a slice that's backed by platform allocated memory for the MTT L2.
+    /// 2: `l1_page_list` is a PageList of 4K pages that must not be aliased elsewhere, and
+    /// must remain allocated until a platform reset.
```

Comment:
Think we are talking past each other regarding the safety requirements. As I see, it:

1. There's no easy way to express the safety requirements for the closure. Basically, we are requiring the closure (the callee) to adhere to the safety requirement for L1 pages, and AFAIK, there's no way to express that in the type system for FnMut(). 
2. Performing the safety check upfront in init() removes the later requirement for other functions like `l1_slice_from_pfn`  to be unsafe because the entries in the tables are either already populated by platform FW, or have been created by Salus at runtime. In either case, the validating for these pages is enforced by init()
3. Another (arguably minor) advantage of performing the allocation upfront is that we don't have to worry about expressing constraints that the callback must not perform any MTT related operations. This is because the callback will be called with an exclusive lock on the MTT tables.

Anyway, it's sort this offline on Monday.

Reply:
Perhaps, your intent is to mark `pub fn set_confidential` and `pub fn set_non_confidential` as unsafe, and express the requirements for the callback as part of that contract.

Reply:
`Page<InternalClean>` is an "owning reference" to a page of memory. In other words, it confers unique ownership of the referenced page of memory to the holder. So where is the unsafety in turning it into an MTT L1 page?

> Performing the safety check upfront in init() removes the later requirement for other functions like l1_slice_from_pfn to be unsafe because the entries in the tables are either already populated by platform FW, or have been created by Salus at runtime. In either case, the validating for these pages is enforced by init()

The `l1_slice_from_pfn` safety issue is orthogonal. That function is effectively dereferencing a `SupervisorPfn`, a type that, unlike `Page<>`, does not imply anything about the ownership or validity of the referenced memory.

Reply:
Please see the updates.

1. `set_confidential` and `set_non_confidential` have been changed to unsafe to allow expression of the constraints on the call back
2. `l1_slice_from_pfn`  is unsafe as well.